### PR TITLE
Update SnackBar to support Material 3

### LINKED ISSUE #110402: Add support for MaterialStateColor for SnackBarAction textColor
As stated by @HansMuller in https://github.com/flutter/flutter/issues/110338#issuecomment-1228642802_
it would be good to add support for `MaterialStateColor` for specifying `textColor` in `SnackBarAction`.

### LINKED ISSUE #114091: Update Snackbar to support Material 3
As part of #91605, we need to migrate the `Snackbar` widget to [Material 3](https://m3.material.io/components/snackbar/overview): 

Screen Shot 2022-10-26 at 11 52 11 AM

### PR DESCRIPTION
**Pull request by esouthren**
Fixes https://github.com/flutter/flutter/issues/114091 (M3 Migration)
Fixes https://github.com/flutter/flutter/issues/110402 (Bring MaterialStateColor to action text)

Updates `SnackBar` to Material 3 specifications. 

https://user-images.githubusercontent.com/15033982/203043484-af7ea16d-1f3b-4632-b631-f4b0ab2bed4a.mov

Specs: https://m3.material.io/components/snackbar/specs
New goldens: https://flutter-gold.skia.org/search?issue=115750&crs=github&patchsets=13&corpus=flutter

This PR adds a few things: 
- Templated M3 file for token values
- Optional `Icon` functionality to SnackBars 
- New Flutter example using Material 3 specs (see video above for screencast)
- Fixes `snackbar.json`, which was referencing `OnInverseSurface` for color scheme values, but should have been `inverseOnSurface`.
- Adds tests to `SnackBar` and `SnackBarTheme` for new icon properties.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: dev/tools/gen_defaults/data/snackbar.json
Comment (darrenaustin) [inline dev/tools/gen_defaults/data/snackbar.json:30]: I didn't realize these were out of sync. I will look into updating the token extractor script to fix it at that end.

File: dev/tools/gen_defaults/lib/snackbar_template.dart
Comment (darrenaustin) [inline dev/tools/gen_defaults/lib/snackbar_template.dart:63]: _iconColor should probably be inlined here as pulling it out doesn't really buy us much.

Comment (darrenaustin) [inline dev/tools/gen_defaults/lib/snackbar_template.dart:63]:
```suggestion
  @override
  Icon get icon =>
      Icon(Icons.close,
        size:  ${tokens["$tokenGroup.icon.size"]},
        color: _iconColor(),
```

File: examples/api/lib/material/snack_bar/snack_bar.2.dart
Comment (darrenaustin) [inline examples/api/lib/material/snack_bar/snack_bar.2.dart:24]:
```suggestion
      theme: ThemeData(eMaterial3: true),
```

File: examples/api/lib/material/snack_bar/snack_bar.2.dart
Comment (darrenaustin) [inline examples/api/lib/material/snack_bar/snack_bar.2.dart:35]: Is this used anywhere? I see `SnackBarBehavior` used below, but I don't see `SnackBarType`.

File: packages/flutter/lib/src/material/snack_bar_theme.dart
Comment (darrenaustin) [inline packages/flutter/lib/src/material/snack_bar_theme.dart:123]: Perhaps this should be just a `Widget?` which would be a little more flexible (i.e. they could wrap an `Icon` with some padding or some such).

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (darrenaustin) [inline packages/flutter/lib/src/material/snack_bar.dart:98]: Nit: unnecessary spacing change.

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (darrenaustin) [inline packages/flutter/lib/src/material/snack_bar.dart:105]: Not sure how to word this better, but perhaps:

```suggestion
  /// If [textColor] is a [MaterialStateColor], then the text color will be 
  /// be resolved against the set of [MaterialState]s that the action
  /// text is in, thus allowing for different colors for pressed, hovered, etc.
  ```

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (darrenaustin) [inline packages/flutter/lib/src/material/snack_bar.dart:157]: Can these also be `MaterialStateColor`? If not, it is a little weird that we can use a `MaterialStateColor` for the widget parameter, but not the theme.

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (darrenaustin) [inline packages/flutter/lib/src/material/snack_bar.dart:186]: Does this need to be public? Why do we need a whole new widget for the close icon?

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (darrenaustin) [inline packages/flutter/lib/src/material/snack_bar.dart:550]: Probably not critical, but we shouldn't create the inverse theme if we don't need it. Perhaps something like:

```dart
final ThemeData effectiveTheme = theme.useMaterial3
  ? theme
  : theme.copyWith(
      colorScheme: // ... inverse color scheme for M2.
```

And then use `effectiveTheme` below when building the Material widget.

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (darrenaustin) [inline packages/flutter/lib/src/material/snack_bar.dart:469]: Nit: unnecessary format changes.

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (darrenaustin) [inline packages/flutter/lib/src/material/snack_bar.dart:592]: Nit: unnecessary format changes.

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (darrenaustin) [inline packages/flutter/lib/src/material/snack_bar.dart:756]:
```suggestion
  final BuildContext context;
  late final ThemeData _theme = Theme.of(context);
  late final ColorScheme _colors = _theme.colorScheme;
  late final TextTheme _textTheme = theme.textTheme;
```

Making these late will allow them to only be computed in the case when they are needed.

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (darrenaustin) [inline packages/flutter/lib/src/material/snack_bar.dart:759]: Given that this does not need to be computed, it could just be given in the constructor's super clause.

File: dev/tools/gen_defaults/lib/snackbar_template.dart
Comment (darrenaustin) [inline dev/tools/gen_defaults/lib/snackbar_template.dart:18]: We can make this a little more efficient if we cache the Theme.of(context) calls below by adding these late variables:

```dart
late final ThemeData _theme = Theme.of(context);
late final ColorScheme _colors = _theme.colorScheme;
```

And then in the constructor set the color prefix to:

```dart
  const SnackbarTemplate(super.blockName, super.fileName, super.tokens, {
    super.colorSchemePrefix = '_colors.'
  });
```

You can also add a similar one for `textTheme` if you are doing a lot of text theme look ups.

**Review by darrenaustin — CHANGES_REQUESTED** [review]
Looks good. Thanks so much for taking this on @esouthren. Comments, suggestions and nits below.

File: dev/tools/gen_defaults/lib/snackbar_template.dart
Comment (esouthren) [reply dev/tools/gen_defaults/lib/snackbar_template.dart:63]: Done

Comment (esouthren) [reply dev/tools/gen_defaults/lib/snackbar_template.dart:63]: Done

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (esouthren) [reply packages/flutter/lib/src/material/snack_bar.dart:756]: Nice - thanks

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (esouthren) [reply packages/flutter/lib/src/material/snack_bar.dart:759]: Done

File: dev/tools/gen_defaults/data/snackbar.json
Comment (esouthren) [reply dev/tools/gen_defaults/data/snackbar.json:30]: Great, thanks - didn't know there was a token script, so that's useful to know!

File: examples/api/lib/material/snack_bar/snack_bar.2.dart
Comment (esouthren) [reply examples/api/lib/material/snack_bar/snack_bar.2.dart:24]: Done

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (esouthren) [reply packages/flutter/lib/src/material/snack_bar.dart:469]: Done

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (esouthren) [reply packages/flutter/lib/src/material/snack_bar.dart:105]: Done - tweaked the ending a little

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (esouthren) [reply packages/flutter/lib/src/material/snack_bar.dart:550]: Good comment, thank you

File: dev/tools/gen_defaults/lib/snackbar_template.dart
Comment (esouthren) [reply dev/tools/gen_defaults/lib/snackbar_template.dart:18]: Done, though I have to remove the `cosnt` constructor of SnackBarDefaultsM3. I don't think there's a way around that?

Comment (darrenaustin) [reply dev/tools/gen_defaults/lib/snackbar_template.dart:18]: I don't think so, but there is never going to be a const with a `BuildContext` here anyway, so I think the trade off is worth it. Thx.

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (esouthren) [reply packages/flutter/lib/src/material/snack_bar.dart:157]: I'm unsure what you mean - do you mean changing the parameter type? If so, wouldn't that be a breaking change?

**Comment by esouthren** [comment]
> Also, I forgot in my review: can you update the docs for `ThemeData.useMaterial3` to include the `SnackBar`? Thx.

Thank you for the reminder - done

**Comment by esouthren** [comment]
Cheers for the swift review Darren, really appreciate it. Couple of unresolved comments/queries from me.

**Comment by LasseRosenow** [comment]
Shouldn't M3 Snackbar always be floating?

**Comment by guidezpl** [comment]
Generally LGTM. A few points

1. > Shouldn't M3 Snackbar always be floating?

  Agree with this, see below for tablet/desktop snackbars.

image

2. Additionally, the insets seem too small at the moment.

3. How are the additional configurations handled?

image

**Comment by esouthren** [comment]
re. 1. floating snackbars, I'm taking my cue from this - 

image

However, this does leave it to the user to enforce snackbars not appearing in front of navigation/app bars, if they use them.

For 2 + 3, let me test and get back to you.

**Comment by esouthren** [comment]
For the additional configurations, I've expanded the Example to include more customisation options: 

https://user-images.githubusercontent.com/15033982/203380278-0eece9c3-8e4e-4a7b-8cfe-beb86d167351.mov

Configurations 1-4 were already handled by the current code. Config 5 (multi line text, with long action) was a little more complex and I can see two solutions - 

1) creating some sort of alignment property on the SnackBarAction widget
2) Calculating an overflow value and pushing the Action to a new line if required. 

I went with (2), as it creates a more dynamic response to the size of the screen and width of the snackbar. However, I'm concerned that this might be a breaking change. 

If the implementation looks good to you, I'll return and add tests to the new code.

**Comment by esouthren** [comment]
> The fade in (and fade out?) animation of the floating snack bar seems to be missing. At least in the video it seems like it just appears without any transition. But in the spec there would be a subtle animation

It has an animation - here's a 10.0 value timeDilation to show it slowed down. 

https://user-images.githubusercontent.com/15033982/203414765-30ca8a15-9987-42dc-b275-fc56d1a31f33.mov

**Comment by LasseRosenow** [comment]
Oh okay then the animation is really short. 

Btw. the animation is very different to how it is on the spec website.

E.g.: https://firebasestorage.googleapis.com/v0/b/design-spec/o/projects%2Fm3%2Fimages%2Fl7qbphkw-3p-snackbars-behavior-2.mp4?alt=media&token=f1814c36-f32a-498b-a456-763b5aa470f1

In general what I find is, that:
- The Text should not face from zero opacity to 1, but from something like 0.5? Could also just seem like that because of the next point:
- The snackbar would expand from bottom to top. Meaning it starts with a height of 0 and then expands to its full height.
- It seems the duration is around 0.5s? (Not entirely sure though)

**Comment by esouthren** [comment]
> 

Nice catch, thank you - I should be able to find the formal animation spec, and can update it from there.

**Comment by esouthren** [comment]
> 
> Nice catch, thank you - I should be able to find the formal animation spec, and can update it from there.

I've partially managed this; I've updated the animation itself so it expands correctly from bottom to top. 

However, I ran into issues trying to modify the animation duration (it's different between M2 and M3). This is because the animation controller is set by the Scaffold, using SnackBar's static method `createAnimationController`. I can't seem to access the correct build context which contains the current value for `ThemeData.useMaterial3` from the Scaffold, in order to conditionally change the animation duration. I've raised [#115924](https://github.com/flutter/flutter/issues/115924) and will triage with the team - there may be a slight delay due to the current holidays.

Normal speed:

https://user-images.githubusercontent.com/15033982/203562183-64d17f46-48b8-4d79-a7dc-7165df8edf93.mov


10% speed:


https://user-images.githubusercontent.com/15033982/203562196-87fc5c97-6795-4709-b4e5-3bae9f015e45.mov

**Comment by LasseRosenow** [comment]
Already much better. I don't know, if this is because of the video, but it still seems a bit different to the spec.

The first frame of your animation already starts with the snackbar being almost full height. (I analyzed the 10% speed version):
| First Frame | Full Snackbar|
| - | - |
| image | image |


In the spec the first visible frame is only a super small part of the snackbar and then it would super smoothly increase its height:
| First Frame | Full Snackbar|
| - | - |
|  image | image |

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (guidezpl) [reply packages/flutter/lib/src/material/snack_bar.dart:180]: I don't think the icon is meant to be interactive, it is?

File: dev/tools/gen_defaults/data/snackbar.json
Comment (darrenaustin) [inline dev/tools/gen_defaults/data/snackbar.json:19]: If you rebase to master, these changes should be there now (see #115890).

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (darrenaustin) [reply packages/flutter/lib/src/material/snack_bar.dart:151]: Sorry, what I was referring to is that we only look at widget.textColor to see if it is a `MaterialStateColor` and resolve it. We should probably do the same for snackBarTheme.actionTextColor and defaults.actionTextColor (i.e. check to see if they are a `MaterialStateColor` and resolve them the same way).

**Review by darrenaustin — CHANGES_REQUESTED** [review]
Looks good. Just a few more issues to resolve. Thx.!

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (esouthren) [reply packages/flutter/lib/src/material/snack_bar.dart:180]: In the spec it's mentioned as a 'Close Icon' but doesn't specify interactivity. Would you propose making it non-customizable, and only have the closing functionality available?

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (esouthren) [reply packages/flutter/lib/src/material/snack_bar.dart:151]: Gotcha - thank you for the explanation!

**Comment by esouthren** [comment]
> Especially also in your current implementation the first visible frame already shows the whole text.
> 
> On the spec you can see no text. because the text is much higher up and thus not visible of course

Thanks @LasseRosenow, I think this is looking much better now. 


https://user-images.githubusercontent.com/15033982/204521430-c2c5154a-250b-43a9-8a23-f82ddcf92e7e.mov

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (guidezpl) [reply packages/flutter/lib/src/material/snack_bar.dart:191]: > Would you propose making it non-customizable, and only have the closing functionality available?

Yes, maybe a `showCloseIcon` flag or similar.

File: packages/flutter/lib/src/material/snack_bar_theme.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/snack_bar_theme.dart:126]: Docs for `SnackBar.margin` and `behavior` should be updated accordingly

File: packages/flutter/lib/src/material/snack_bar_theme.dart
Comment (guidezpl) [inline packages/flutter/lib/src/material/snack_bar_theme.dart:65]: the biggest nit of all time, feel free to ack, but I'd switch the order so that `insetPadding` is next to `width`

**Review by guidezpl — CHANGES_REQUESTED** [review]
Very close!

File: packages/flutter/lib/src/material/snack_bar_theme.dart
Comment (esouthren) [reply packages/flutter/lib/src/material/snack_bar_theme.dart:65]: Ha, no problem. Makes sense to group them together!

File: packages/flutter/lib/src/material/snack_bar_theme.dart
Comment (esouthren) [reply packages/flutter/lib/src/material/snack_bar_theme.dart:126]: Done, thanks

File: packages/flutter/lib/src/material/snack_bar.dart
Comment (esouthren) [reply packages/flutter/lib/src/material/snack_bar.dart:191]: Sure - that simplifies things. Removed the SnackBarIcon class, and updated the tests accordingly.

**Review by guidezpl — APPROVED** [review]
```
  _    ___ _____ __  __ 
 | |  / __|_   _|  \/  |
 | |_| (_ | | | | |\/| |
 |____\___| |_| |_|  |_|
                        
```

**Review by darrenaustin — APPROVED** [review]
flutter-lgtm

Nice work @esouthren!

**Comment by Piinks** [comment]
This PR was merged without handling the golden file images linked above and has broken the tree. @esouthren can you view the images here and approve them if they look correct? Otherwise we will need to revert the change to reopen the tree.

**Comment by Piinks** [comment]
I don't know if the images can be approved or not, so we'll need to revert this for now until they can be addressed.

**Pull request merged**
